framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,531 @@
+    }
+
+    /**
+     * Prints a literal; used when printing the arguments of annotations. Nearly a copy of the
+     * same-named method in {@code IndexFileWriter}, with one modification: this version (correctly)
+     * prints long literals with an L at the end, so that they are valid Java source code (if they
+     * are larger than Integer.MAX_VALUE).
+     *
+     * @param aft the annotation in which we are printing
+     * @param o the value or values to print
+     */
+    private void printValue(AnnotationFieldType aft, Object o) {
+        if (aft instanceof AnnotationAFT) {
+            printAnnotation((Annotation) o);
+        } else if (aft instanceof ArrayAFT) {
+            ArrayAFT aaft = (ArrayAFT) aft;
+            printWriter.print('{');
+            if (!(o instanceof List)) {
+                printValue(aaft.elementType, o);
+            } else {
+                List<?> l = (List<?>) o;
+                // watch out--could be an empty array of unknown type
+                // (see AnnotationBuilder#addEmptyArrayField)
+                if (aaft.elementType == null) {
+                    if (l.size() != 0) {
+                        throw new AssertionError();
+                    }
+                } else {
+                    boolean first = true;
+                    for (Object o2 : l) {
+                        if (!first) {
+                            printWriter.print(',');
+                        }
+                        printValue(aaft.elementType, o2);
+                        first = false;
+                    }
+                }
+            }
+            printWriter.print('}');
+        } else if (aft instanceof ClassTokenAFT) {
+            printWriter.print(aft.format(o));
+        } else if (aft instanceof BasicAFT && o instanceof String) {
+            printWriter.print(Strings.escape((String) o));
+        } else if (aft instanceof BasicAFT && o instanceof Long) {
+            printWriter.print(o.toString() + "L");
+        } else {
+            printWriter.print(o.toString());
+        }
+    }
+
+    /**
+     * Prints an annotation in Java source format.
+     *
+     * @param a the annotation to print
+     */
+    private void printAnnotation(Annotation a) {
+        printWriter.print("@" + a.def().name.substring(a.def().name.lastIndexOf('.') + 1));
+        if (!a.fieldValues.isEmpty()) {
+            printWriter.print('(');
+            boolean first = true;
+            for (Map.Entry<String, Object> f : a.fieldValues.entrySet()) {
+                if (!first) {
+                    printWriter.print(',');
+                }
+                printWriter.print(f.getKey() + "=");
+                printValue(a.def().fieldTypes.get(f.getKey()), f.getValue());
+                first = false;
+            }
+            printWriter.print(')');
+        }
+    }
+
+    /**
+     * Prints all annotations in {@code annos}, separated by spaces. See {@link #printAnnotation}.
```

Comment:
"separated by spaces" should be "each preceded by a space".  But, I suggest changing the logic so each is followed by a space.  Otherwise, when it is called just after printing INDENT, the result is inconsistent indentation.